core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -59,24 +61,101 @@ object RequestChannel extends Logging {
+    val consumerFetchRequestSizeMetricNameMap = getConsumerFetchRequestAcksSizeMetricNameMap
+
     (enabledApis.map(_.name) ++
       Seq(RequestMetrics.MetadataAllTopics) ++
-      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName)).foreach { name =>
-      metricsMap.put(name, new RequestMetrics(name))
+      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName) ++
+      getConsumerFetchRequestSizeMetricNames ++
+      getProduceRequestAcksSizeMetricNames)
+      .foreach { name => metricsMap.put(name, new RequestMetrics(name))
     }
 
     def apply(metricName: String): RequestMetrics = metricsMap(metricName)
 
     def close(): Unit = {
        metricsMap.values.foreach(_.removeMetrics())
     }
+
+    // generate map[request/responseSize, metricName] for requests of different size
+    def getRequestSizeMetricNameMap(requestType: String):Map[Int, String] = {
+      // convert requestMetricsSizeBucketsConfig from string to int array. e.g., "0,1,10,50,100" => [0,1,10,50,100]
+      var bucketsString = requestMetricsSizeBucketsConfig.replaceAll("\\s", "").split(',')
+      if(bucketsString.length < 2) {
+        error(s"requestMetricsSizeBuckets has fewer than 2 buckets: '${requestMetricsSizeBucketsConfig}'. " +
+          s"use default value instead ")
+        bucketsString = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',')
+      }
+      var bucketsInt = Array[Int](bucketsString.length)
+      try {
+        bucketsInt = bucketsString.map(_.toInt)
+      } catch {
+        case e: Exception => {
+          error(s"Failed to parse requestMetricsSizeBuckets: '${requestMetricsSizeBucketsConfig}'. " +
+            s"use default value instead ", e)
+          bucketsInt = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',').map(_.toInt)
+        }
+      }
+
+      // get size and corresponding term to be used in metric name
+      // [0,1,10,50,100] => requestSizeBuckets:Seq((0, "0To1Mb"), (1, "1To10Mb"), (10, "10To50Mb"), (50, "50To100Mb"),
+      // (100, "100MbGreater"))
+      val requestSizeBuckets = for (i <- 0 until bucketsInt.length) yield {
+        val size = bucketsInt(i)
+        if (i == bucketsInt.length - 1) (size, s"${size}MbGreater")
+        else (size, s"${size}To${bucketsInt(i + 1)}Mb")
+      }
+      requestSizeBuckets.map(bucket => (bucket._1, s"${requestType}${bucket._2}")).toMap
+    }
+
+    // generate map[acks, map[requestSize, metricName]] for produce requests of different request size and acks
+    def getProduceRequestAcksSizeMetricNameMap():Map[Int, Map[Int, String]] = {
+      val produceRequestAcks = Seq((0, "0"), (1, "1"), (-1, "All"))
+      val requestSizeMetricNameMap = getRequestSizeMetricNameMap(ApiKeys.PRODUCE.name)
+
+      produceRequestAcks.map(ack =>
+        (ack._1, requestSizeMetricNameMap.map({case(size, name) => (size, s"${name}Acks${ack._2}")}))).toMap
```

Comment:
Maybe add a delimiter between the name and Acks.

Reply:
Prefer it like this "Produce20To200MbAcks0" with delimiter. It is readable. Our ingraph for latency has name likes "RequestStats.FetchConsumer-TotalTimeMs.999thPercentile" which uses "-" to separate sensor name (FetchConsumer) and metric name (TotalTime...). If adding "_" to sensor name, I feel it is not very straightforward to consider "Produce_20To200Mb_Acks0" as a whole term to present sensor name.

Reply:
Ok, thanks for the explanation.